File: Example/PDStrategy/PDStrategy/PDStrategy/Models/PDItem.m
//
//  PDItem.m
//  PDStrategy
//
//  Created by Pavel Deminov on 19/11/2017.
//  Copyright © 2017 Pavel Deminov. All rights reserved.
//

#import "PDItem.h"

@implementation PDItem

+ (nonnull PDItem *)instantiate {
    PDItem *item = [PDItem new];
    return item;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        [self setup];
    }
    return self;
}

- (void)setup {
    
}

- (void)validate {
    [self invalidate];
    ValidationState state = ValidationStateValid;
    
    for (PDRule *rule in self.rules) {
        BOOL valid = [rule validate:self.value];
        if (!valid) {
            self.errorRule = rule;
            state = ValidationStateInvalid;
            break;
        }
    }
    self.state = state;
}

- (void)invalidate {
    self.state = ValidationStateNone;
    self.errorRule = nil;
}

- (UpdateBlock)addUpdateBlock:(UpdateBlock)updateBlock {
    if (!self.updateBlocks) {
        self.updateBlocks = [NSMutableArray new];
    }
    [self.updateBlocks addObject:updateBlock];
    return updateBlock;
}

- (void)removeUpdateBlock:(UpdateBlock)updateBlock {
    [self.updateBlocks removeObject:updateBlock];
}

#pragma mark PDItemInfo

@synthesize pdApiKey;

@synthesize pdDate;

@synthesize pdImage;

@synthesize pdImageUrl;

@synthesize pdItemHash;

@synthesize pdLocked;

@synthesize pdPlaceholder;

@synthesize pdTitle;

@synthesize pdType;

- (NSString *)pdTitle {
    return self.title;
}

- (id)pdValue {
    return self.value;
}

- (void)setPdValue:(id)pdValue {
    self.value = pdValue;
}

- (NSString *)pdItemHash {
    return self.itemHash;
}

- (NSDate *)pdDate {
    return self.date;
}

- (NSString *)pdApiKey {
    return self.apiKey;
}

- (NSString *)pdPlaceholder {
    return self.placeholder;
}

- (id)pdType {
    return self.type;
}

- (UIImage *)pdImage {
    return self.image;
}

- (NSURL *)pdImageUrl {
    return self.imageUrl;
}

- (BOOL)pdLocked {
    return self.locked;
}

@end

File: Example/PDStrategy/PDStrategy/PDStrategy/Models/PDItem.swift
//
//  PDItem.swift
//  PDStrategy
//
//  Created by Pavel Deminov on 04/11/2017.
//  Copyright © 2017 Pavel Deminov. All rights reserved.
//

import UIKit

enum PDValidationState  {
    case none
    case valid
    case invalid
}

class PDItem: NSObject, PDItemInfo {
    var itemHash: String? {
        get {
            return String(hash)
        }
    }
    
    var title: String?
    var value: Any?
    var date: Date?
    var apiKey: String?
    var placeholder: String?
    var type: Any?
    var image: UIImage?
    var imageUrl: URL?
    var apiRequred: Bool?
    
    var validationState = PDValidationState.none
    var errorRule: PDRule?
    var rules: [PDRule]?
    
    class func item() -> PDItem {
        let item = PDItem()
        item.setup()
        return item
    }
    
    func setup() {
        
    }
    
    func validate() {
        invalidate()
        validationState = .valid
        guard let rules = rules else {
            return
        }
        
        for rule: PDRule in rules {
            let valid: Bool = rule.validate(value as Any)
            if !valid {
                errorRule = rule
                validationState = .invalid
                break
            }
        }
    }
    
    func invalidate() {
        validationState = PDValidationState.none
        errorRule = nil
    }
}
